### PR TITLE
Support multiple url factories, fix issue with createUrl

### PR DESCRIPTION
**Pull request by RdeWilde**
We had issues with createUrl when factory was not yet initialized. Also we added support for saving multiple factories.
Was not extensively tested yet.

### CODE REVIEW

File: src/UrlRule.php [hunk 1]
```diff
@@ -167,29 +224,24 @@ public function parseRequest($manager, $request)
...
+            if (YII_DEBUG) {
+            	(new $this->modelClass)->getRelation($relation);
+			}
 
             $modulePrefix = isset($this->modulePrefix) ? $this->modulePrefix .'/' : '';
             $this->config['controller'][$urlName] = $modulePrefix . $controller;
 
-            $this->setRulesFactory($this->config);
-            $routeObj = $this->rulesFactory->parseRequest($manager, $request);
+			$factory = Yii::createObject($this->config);
```
Comment (RdeWilde) [inline src/UrlRule.php:242]: Add; Check if already exists?

File: src/UrlRule.php [hunk 2]
```diff
@@ -8,6 +8,7 @@
 namespace tunecino\nestedrest;
 
 use Yii;
+use yii\helpers\VarDumper;
```
Comment (RdeWilde) [inline src/UrlRule.php:11]: Leftover

**Comment by tunecino** [comment]
Thanks for contributing. I'll have a look at it this weekend. by the meantime could you please provide more info about the issue you did encounter. how does your rules configuration look like ? and what where the expected results or the expected uri endpoints ?

**Comment by RdeWilde** [comment]
Kind a complicated, but it triggered createUrl before parseRequest was triggered (so no factory existed) - I think it was due the getLinks() functions for rest Links.

Sometimes this was fixed by placing the 'normal' yii REST rule above the nested one, but for particular cases that didn't do the trick.

This PR is purely to share our progress and changes with you. Feel free to use what you can and leave what you won't :)

**Comment by tunecino** [comment]
I couldn't figure out what may had caused that issue.

Generally speaking default **yii\rest\UrlRule** was designed to receive a config array holding a list of controllers and related options like verbs, actions, etc. For each of those controllers as shown [here](https://github.com/yiisoft/yii2/blob/e32f2f4c092e424794c55fe3cbe9598e0669f97f/framework/rest/UrlRule.php#L175) a **yii\web\UrlRule** related config array will be dynamically generated then used to create all related **yii\web\UrlRule** objects. Parsing the request then will be comparing the requested url against that list of **web** rules until finding the first matching **controller/action** path to which the request should be redirected.

To manually define nested routing in Yii you'll need a similar configs to [those](https://github.com/tunecino/Yii2_foundation-apps/blob/master/backend/api/config/rules.php#L30-L50). Wich will be parsed one by one until getting that first **controller/action** matching to do the redirection as shown [here](https://github.com/yiisoft/yii2/blob/e32f2f4c092e424794c55fe3cbe9598e0669f97f/framework/rest/UrlRule.php#L222-L226), the next defined rules then will simply be ignored. Which looks correct to me at least from a performance point of view as only one of the defined rules is supposed to satisfy the requested url.

Now back to this extension, the role of its custom UrlRule class is to dynamically generate those **yii\rest\UrlRule** config arrays, and **one by one**, creates its related **yii\rest\UrlRule** object and see if it leads to a valid **controller/action** matching path within 

``` php
$routeObj = $this->rulesFactory->parseRequest($manager, $request);
```

before going to create and parse the next rule config array. and that is why I can't see how generating multiple instances of `yii\rest\UrlRule::parseRequest()` at once and duplicating arrays could lead to a different result.

Anyway. I think this PR should remain open as I need to write tests for the extension covering all cases first. Also I need to find that case triggering the **createUrl** method before **parseRequest**. By the meantime someone else may also get a similar issue so your suggestion may help solving his. and he may help us understanding better what was going wrong with current code. so thanks for sharing.

**Comment by yairpe** [comment]
I had a similar issue about the missing factory

In some places, i used to get the absolute URL with

`Yii::$app->urlManager->createAbsoluteUrl('')`

And it complained:

```
"name": "Exception",
    "message": "Call to a member function createUrl() on null",
    "code": 0,
    "type": "Error",
    "file": "C:\\Apache24\\htdocs\\apps\\topihub\\vendor\\tunecino\\yii2-nested-rest\\src\\UrlRule.php",
    "line": 146,
```

because the `return $this->rulesFactory->createUrl($manager, $route, $params);` has no factory 

I changed my code to use `Url::home(true)` to bypass it.

Maybe it'll help you

**Comment by tunecino** [comment]
I guess it is raised when`yii/web/UrlManager` tries to evaluate all the `createUrl()` functions of all the registered `rules` [here](https://github.com/yiisoft/yii2/blob/74068eb05378e52222bb2721de5fcb1c294c0270/framework/web/UrlManager.php#L376-L389) even those not being initialized. It should be fixed by https://github.com/tunecino/yii2-nested-rest/commit/b870f61dbd7a3f53753c40d663bbd33a4283fc44. 
@yairpe could you please update to [0.2.1](https://github.com/tunecino/yii2-nested-rest/releases/tag/0.2.1) and see if it works for your case?

**Comment by yairpe** [comment]
It is ok now. Thanks

**Comment by tunecino** [comment]
Great. I'll close this PR then as it shouldn't be needed after the fix. Thanks to both of you.

**Pull request closed without merging**